File: FHD/app/schemas/tutorial_v2.py
"""Public DTO contract for Tutorial V2."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class TutorialEvidenceDTO(BaseModel):
    step_id: str
    status: Literal["pending", "failed", "passed"]
    result_code: str
    entity_refs: list[dict[str, Any]] = Field(default_factory=list)
    counts: dict[str, int | float | str] = Field(default_factory=dict)
    attempt_count: int = 0
    verified_at: str | None = None


class TutorialStepDTO(BaseModel):
    id: str
    title: str
    goal: str
    instruction: str
    success_criteria: str
    why: str
    hint: str
    route_name: str
    target_selector: str
    required: bool = True
    status: Literal["pending", "failed", "passed"] = "pending"
    evidence: TutorialEvidenceDTO | None = None


class TutorialRunDTO(BaseModel):
    id: str
    workspace_id: str
    course_id: str
    version: int
    status: Literal["active", "paused", "completed", "reset"]
    current_step_id: str
    attempt_count: int
    progress: int = Field(ge=0, le=100)
    completed_steps: int
    total_steps: int
    generation: int
    teaching_space: Literal[True]
    steps: list[TutorialStepDTO]
    started_at: str | None = None
    completed_at: str | None = None


class TutorialCourseDTO(BaseModel):
    id: str
    title: str
    summary: str
    estimated_minutes: int
    prerequisite_ids: list[str]
    version: int
    steps: list[TutorialStepDTO]
    locked: bool
    missing_prerequisite_ids: list[str]
    run: TutorialRunDTO | None = None
    status: Literal["not_started", "active", "paused", "completed", "reset"]
    progress: int = Field(ge=0, le=100)


__all__ = [
    "TutorialCourseDTO",
    "TutorialEvidenceDTO",
    "TutorialRunDTO",
    "TutorialStepDTO",
]

File: FHD/frontend/src/api/tutorialV2.ts
import api from './core'

export interface TutorialEvidenceDTO {
  step_id: string
  status: 'pending' | 'failed' | 'passed'
  result_code: string
  entity_refs: Array<{ type: string; id: string | number }>
  counts: Record<string, string | number>
  attempt_count: number
  verified_at: string | null
}

export interface TutorialStepDTO {
  id: string
  title: string
  goal: string
  instruction: string
  success_criteria: string
  why: string
  hint: string
  route_name: string
  target_selector: string
  required: boolean
  status: 'pending' | 'failed' | 'passed'
  evidence: TutorialEvidenceDTO | null
}

export interface TutorialRunDTO {
  id: string
  workspace_id: string
  course_id: string
  version: number
  status: 'active' | 'paused' | 'completed' | 'reset'
  current_step_id: string
  attempt_count: number
  progress: number
  completed_steps: number
  total_steps: number
  generation: number
  teaching_space: true
  steps: TutorialStepDTO[]
  started_at: string | null
  completed_at: string | null
}

export interface TutorialCourseDTO {
  id: string
  title: string
  summary: string
  estimated_minutes: number
  prerequisite_ids: string[]
  version: number
  steps: TutorialStepDTO[]
  locked: boolean
  missing_prerequisite_ids: string[]
  run: TutorialRunDTO | null
  status: 'not_started' | TutorialRunDTO['status']
  progress: number
}

type ApiEnvelope<T> = { success: boolean; data: T }

export const tutorialV2Api = {
  async courses(): Promise<TutorialCourseDTO[]> {
    const response = await api.get<ApiEnvelope<TutorialCourseDTO[]>>('/api/tutorial/v2/courses')
    return response.data
  },
  async start(courseId: string): Promise<TutorialRunDTO> {
    const response = await api.post<ApiEnvelope<TutorialRunDTO>>('/api/tutorial/v2/runs', {
      course_id: courseId,
    })
    return response.data
  },
  async current(): Promise<TutorialRunDTO | null> {
    const response = await api.get<ApiEnvelope<TutorialRunDTO | null>>('/api/tutorial/v2/runs/current')
    return response.data
  },
  async enter(runId: string): Promise<TutorialRunDTO> {
    const response = await api.post<ApiEnvelope<TutorialRunDTO>>(`/api/tutorial/v2/runs/${runId}/enter`)
    return response.data
  },
  async leave(runId: string): Promise<TutorialRunDTO> {
    const response = await api.post<ApiEnvelope<TutorialRunDTO>>(`/api/tutorial/v2/runs/${runId}/leave`)
    return response.data
  },
  async verify(runId: string, stepId: string, context: Record<string, unknown>) {
    const response = await api.post<ApiEnvelope<{
      run: TutorialRunDTO
      evidence: TutorialEvidenceDTO
      hint: string
    }>>(`/api/tutorial/v2/runs/${runId}/steps/${stepId}/verify`, { context })
    return response.data
  },
  async reset(runId: string): Promise<TutorialRunDTO> {
    const response = await api.post<ApiEnvelope<TutorialRunDTO>>(`/api/tutorial/v2/runs/${runId}/reset`)
    return response.data
  },
  async reports(): Promise<Array<Record<string, unknown>>> {
    const response = await api.get<ApiEnvelope<Array<Record<string, unknown>>>>('/api/tutorial/v2/reports')
    return response.data
  },
}

export default tutorialV2Api
